test/parallel/test-stream-transform-by.js

```diff
@@ -0,0 +1,249 @@
+'use strict';
+const { mustCall } = require('../common');
+const { once } = require('events');
+const { Readable, Transform } = require('stream');
+const { strictEqual } = require('assert');
+
+async function transformBy() {
+  const readable = Readable.from('test');
+  async function * mapper(source) {
+    for await (const chunk of source) {
+      yield chunk.toUpperCase();
+    }
+  }
+
+  const stream = Transform.by(mapper);
+  readable.pipe(stream);
+  const expected = ['T', 'E', 'S', 'T'];
+  for await (const chunk of stream) {
+    strictEqual(chunk, expected.shift());
+  }
+}
+
+async function transformByFuncReturnsObjectWithSymbolAsyncIterator() {
+  const readable = Readable.from('test');
+  const mapper = (source) => ({
+    [Symbol.asyncIterator]() {
+      return {
+        async next() {
+          const { done, value } = await source.next();
+          return { done, value: value ? value.toUpperCase() : value };
+        }
+      };
+    }
+  });
+
+  const stream = Transform.by(mapper);
+  readable.pipe(stream);
+  const expected = ['T', 'E', 'S', 'T'];
+  for await (const chunk of stream) {
+    strictEqual(chunk, expected.shift());
+  }
+}
+
+async function transformByEncoding() {
+  const readable = Readable.from('test');
+  async function * mapper(source) {
+    for await (const chunk of source) {
+      strictEqual(source.encoding, 'ascii');
+      yield chunk.toUpperCase();
+    }
+  }
+  const stream = Transform.by(mapper);
+  stream.setDefaultEncoding('ascii');
+  readable.pipe(stream);
+
+  const expected = ['T', 'E', 'S', 'T'];
+  for await (const chunk of stream) {
+    strictEqual(chunk, expected.shift());
```

Comment:
Now this is failing:

```
python tools/test.py test/parallel/test-stream-transform-by.js

=== release test-stream-transform-by ===                   
Path: parallel/test-stream-transform-by
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^

AssertionError [ERR_ASSERTION]: Expected values to be strictly equal:

'TEST' !== 'T'

    at Duplex.<anonymous> (/Users/safareli/dev/deepchannel/node/test/parallel/test-stream-transform-by.js:130:5)
    at Duplex.emit (node:events:365:28)
    at addChunk (node:internal/streams/readable:314:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at Duplex.Readable.push (node:internal/streams/readable:228:10)
    at next (node:internal/streams/from:98:31)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  generatedMessage: true,
  code: 'ERR_ASSERTION',
  actual: 'TEST',
  expected: 'T',
  operator: 'strictEqual'
}
Command: out/Release/node /Users/safareli/dev/deepchannel/node/test/parallel/test-stream-transform-by.js
[00:00|% 100|+   0|-   1]: Done   
```


I suspect this was working with the original implementation.

I'm really not sure that's the issue :/

Reply:
This test is wrong. If you want multiple chunks here you'd need to move from `Readable.from('test')` to use an array or split it yourself in multiplie chunks.

Reply:
It's not just this but other tests are doing similar things. `transformBy`, `transformByFuncReturnsObjectWithSymbolAsyncIterator` to name a few (there are total 7 tests like this). And this tests are from the previous PR and I suspect It was passing there.

Reply:
We changed the behavior of `Readable.from()` lately so that strings are not split char-by-char. This test must be changed.

Reply:
@safareli any updates?

Reply:
In the recent commit I've fixed some of the test errors but I'm still getting this which not sure how to fix and I gave up:

```
python tools/test.py test/parallel/test-stream-transform-by.js
[00:00|%   0|+   0|-   0]: release test-stream-trans                                                    === release test-stream-transform-by ===
Path: parallel/test-stream-transform-by
Mismatched noop function calls. Expected exactly 1, actual 0.
    at mustCall (/Users/safareli/dev/deepchannel/node/test/common/index.js:350:10)
    at transformBySourceIteratorCompletes (/Users/safareli/dev/deepchannel/node/test/parallel/test-stream-transform-by.js:64:21)
    at Object.<anonymous> (/Users/safareli/dev/deepchannel/node/test/parallel/test-stream-transform-by.js:240:3)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47
Mismatched <anonymous> function calls. Expected exactly 1, actual 0.
    at mustCall (/Users/safareli/dev/deepchannel/node/test/common/index.js:350:10)
    at transformByOnErrorAndTryCatchAndDestroyed (/Users/safareli/dev/deepchannel/node/test/parallel/test-stream-transform-by.js:206:24)
    at Object.<anonymous> (/Users/safareli/dev/deepchannel/node/test/parallel/test-stream-transform-by.js:247:3)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47
Mismatched <anonymous> function calls. Expected exactly 1, actual 0.
    at mustCall (/Users/safareli/dev/deepchannel/node/test/common/index.js:350:10)
    at transformByThrowPriorToForAwait (/Users/safareli/dev/deepchannel/node/test/parallel/test-stream-transform-by.js:229:22)
    at Object.<anonymous> (/Users/safareli/dev/deepchannel/node/test/parallel/test-stream-transform-by.js:248:3)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47
Mismatched noop function calls. Expected exactly 1, actual 0.
    at mustCall (/Users/safareli/dev/deepchannel/node/test/common/index.js:350:10)
    at Object.<anonymous> (/Users/safareli/dev/deepchannel/node/test/parallel/test-stream-transform-by.js:249:9)
    at Module._compile (node:internal/modules/cjs/loader:1109:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1138:10)
    at Module.load (node:internal/modules/cjs/loader:989:32)
    at Function.Module._load (node:internal/modules/cjs/loader:829:14)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
    at node:internal/main/run_main_module:17:47
Command: out/Release/node /Users/safareli/dev/deepchannel/node/test/parallel/test-stream-transform-by.js
                                                    [00:00|% 100|+   0|-   1]: Done
```